Source/CarthageKit/Xcode.swift

```diff
@@ -189,11 +189,25 @@ extension BuildArguments: CustomStringConvertible {
 public func locateProjectsInDirectory(directoryURL: NSURL) -> SignalProducer<ProjectLocator, CarthageError> {
 	let enumerationOptions: NSDirectoryEnumerationOptions = [ .SkipsHiddenFiles, .SkipsPackageDescendants ]
 
-	return NSFileManager.defaultManager().carthage_enumeratorAtURL(directoryURL.URLByResolvingSymlinksInPath!, includingPropertiesForKeys: [ NSURLTypeIdentifierKey ], options: enumerationOptions, catchErrors: true)
-		.reduce([]) { (matches: [ProjectLocator], tuple) -> [ProjectLocator] in
+	return submodulesInRepository(directoryURL)
+		.map { directoryURL.URLByAppendingPathComponent($0.path) }
+		.concat(SignalProducer<NSURL, CarthageError>(value: directoryURL.URLByAppendingPathComponent("Carthage")))
+		.collect()
+		.flatMap(.Merge) { directoriesToSkip in
+			return NSFileManager.defaultManager().carthage_enumeratorAtURL(directoryURL.URLByResolvingSymlinksInPath!, includingPropertiesForKeys: [ NSURLTypeIdentifierKey ], options: enumerationOptions, catchErrors: true)
+				.filter { _, URL in
```

Comment:
I thought that to filter URLs in this phase is inefficient. We could reduce the evaluations by just filtering out the matched `[ProjectLocator]`. So this func could be written as follows:

```swift
public func locateProjectsInDirectory(directoryURL: NSURL) -> SignalProducer<ProjectLocator, CarthageError> {
	let enumerationOptions: NSDirectoryEnumerationOptions = [ .SkipsHiddenFiles, .SkipsPackageDescendants ]

	return NSFileManager.defaultManager().carthage_enumeratorAtURL(directoryURL.URLByResolvingSymlinksInPath!, includingPropertiesForKeys: [ NSURLTypeIdentifierKey ], options: enumerationOptions, catchErrors: true)
		.reduce([]) { (matches: [ProjectLocator], tuple) -> [ProjectLocator] in
			var matches = matches
			let (_, URL) = tuple

			if let UTI = URL.typeIdentifier.value {
				if (UTTypeConformsTo(UTI, "com.apple.dt.document.workspace")) {
					matches.append(.Workspace(URL))
				} else if (UTTypeConformsTo(UTI, "com.apple.xcode.project")) {
					matches.append(.ProjectFile(URL))
				}
			}

			return matches
		}
		.flatMap(.Merge) { matches in
			return submodulesInRepository(directoryURL)
				.map { directoryURL.URLByAppendingPathComponent($0.path) }
				.concat(SignalProducer(value: directoryURL.URLByAppendingPathComponent("Carthage")))
				.collect()
				.map { directoriesToSkip in
					return matches.filter { match in
						for directory in directoriesToSkip {
							if directory.hasSubdirectory(match.fileURL) {
								return false
							}
						}
						return true
					}
				}
		}
		.map { $0.sort() }
		.flatMap(.Merge) { SignalProducer(values: $0) }
}
```